### PR TITLE
[vulnerability] A newly appended block header should not build on an invalid chain

### PR DESCRIPTION
**Pull request by p-j01**
# Description

Fix suggestion to #2280 raised by @yixiao5428 . Integrated bitcoin/bitcoin@015a525 changes to `src/validation.cpp`.
# Issue Related
#2280

### CODE REVIEW

**Comment by patricklodder** [comment]
@p-j01 Could you please rebase this against the current 1.14.4-dev so that we get CI? Thanks!

**Comment by mmicael1** [comment]
@p-j01 Could you please rebase this against the current 1.14.4-dev? Thanks!

**Comment by patricklodder** [comment]
@p-j01 I think what we need is:

```
git remote add upstream https://github.com/dogecoin/dogecoin.git
git fetch upstream
git rebase -i upstream/1.14.4-dev
git push -f origin 1.14.4-dev
```

When you get into interactive rebase mode, delete all the lines that contain commits that aren't yours, and this will be clean :heart:

File: src/validation.cpp [hunk 1]
```diff
@@ -3153,6 +3191,8 @@ static bool AcceptBlockHeader(const CBlockHeader& block, CValidationState& state
         if (!CheckBlockHeader(block, state))
             return error("%s: Consensus::CheckBlockHeader: %s, %s", __func__, hash.ToString(), FormatStateMessage(state));
 
+        
+
```
Comment (patricklodder) [inline src/validation.cpp:3195]: Nit: these extra lines can be removed

File: src/validation.cpp [hunk 2]
```diff
@@ -3588,6 +3643,10 @@ bool static LoadBlockIndexDB(const CChainParams& chainparams)
             } else {
                 pindex->nChainTx = pindex->nTx;
             }
+            if (!(pindex->nStatus & BLOCK_FAILED_MASK) && pindex->pprev && (pindex->pprev->nStatus & BLOCK_FAILED_MASK)) {
+                pindex->nStatus |= BLOCK_FAILED_CHILD;
+                setDirtyBlockIndex.insert(pindex);
+        }
```
Comment (patricklodder) [inline src/validation.cpp:3649]: Nit: please fix this indentation

File: src/validation.cpp [hunk 3]
```diff
@@ -158,6 +158,26 @@ namespace {
...
+      * to connect a bad block and fail.
+      *
+      * While this is more complicated than marking everything which descends
+      * from an invalid block as invalid at the time we discover it to be
+      * invalid, doing so would require walking all of mapBlockIndex to find all
+      * descendants. Since this case should be very rare, keeping track of all
+      * BLOCK_FAILED_VALID blocks in a set should be just fine and work just as
+      * well.
+      *
+      * Because we alreardy walk mapBlockIndex in height-order at startup, we go
```
Comment (patricklodder) [inline src/validation.cpp:175]: Nit: typo here, alreardy -> already

**Review by patricklodder — CHANGES_REQUESTED** [review]
I like the concept and only have a couple of nits on the proposed code.

Other than that, we would need unit (and perhaps functional, for pruned nodes) tests, at the very least demonstrating a happy scenario so that we can test for regressions in the future.

**Comment by micaelmalta** [comment]
@p-j01 can you change the naming of **_g_failed_blocks_** to be  lower camel case compliant please?

File: src/validation.cpp (hunk 2 above)
Comment (p-j01) [reply src/validation.cpp]: `clang-format` could be used and added to CI for code formatting

**Comment by p-j01** [comment]
`clang-format` could be used and added to CI for code formatting

**Comment by patricklodder** [comment]
I was going to recommend adding `p2p-acceptblock` to standard QA tests (bitcoin did that too), but for some reason this is failing, so I am investigating that before continuing with my review.

**Review by patricklodder — COMMENTED** [review]
Leaving interim comments until the acceptblock tests are fixed.

**Comment by patricklodder** [comment]
@p-j01 There's a non-intrusive way to unblock p2p-acceptblock now with #2448. Please let us know if you're still wanting to work on this. If yes, awesome. If no, I will take over final changes and re-propose this.

**Comment by p-j01** [comment]
@patricklodder If the block index is smaller than 1 or greater than 1439?

**Comment by patricklodder** [comment]
> If the block index is smaller than 1 or greater than 1439?

[Bitcoin Core added 2 steps to p2p-acceptblock to test this patch](https://github.com/bitcoin/bitcoin/commit/5bec7744d112141848e57d7f628dc17597f89dd5). I think @xanimo had already some of this staged on his fork while he was working on fixing #2412 - maybe we can pull either in.

**Comment by ReverseControl** [comment]
@patricklodder What is the best place to start if I want to help with this one?

**Comment by ReverseControl** [comment]
@p-j01 Are you currently working on this?

**Comment by patricklodder** [comment]
This was merged through #2457. Thank you for this - it's much appreciated!

**Pull request closed without merging**